2025-04-27 Enhancements to referring to actions by commit hash.md

```diff
@@ -0,0 +1,149 @@
+
+To solve it, we could provide a way to freeze not only the action's commit has, but also the catalog typing's commit
+hash. However, this problem is neglected for the purpose of this document because demand for this feature is unknown
+(there is no signal about it from the users), and it would lead to further complicating the system.
+
+### Pinning to a specific binding (e.g. by JAR's checksum)
+
+One could imagine an attack on the bindings server where a malicious piece of code is injected to the Kotlin-based
+logic. To address it, we could have a way of specifying e.g. a checksum of the binding's JAR, to be sure that the
+delivered binding is always the same.
+
+However, there's no native support for dependency checksums in Kotlin Scripting, and implementing it in an acceptable
+way from user's point of view is not possible. It would be also challenging to ensure that the JAR provided by the
+server doesn't change - in practice, it's not possible as even a subtle change e.g. in tooling used to build the JAR may
+lead to changing the checksum.
+
+Recommended workarounds are
+* private hosting of the bindings server, either by building it from source (then one can pin the
+  [repo](https://github.com/typesafegithub/github-workflows-kt)'s commit) or pinning to an image's digest in
+  [DockerHub](https://hub.docker.com/r/krzema12/github-workflows-kt-jit-binding-server/tags)
+* fetching the binding to a static location, and consuming it from the script. However, this
+  way makes using dependency updating bots extremely hard
+
+## Proposed solution
+
+Add two new ways of consuming an action binding:
+
+### 1. With validation of (version, commit hash) pair
+
+Example:
+
+```kotlin
+@file:Repository("https://bindings.krzeminski.it")
+@file:DependsOn("actions:checkout__commit:v4.1.2__85e6279cec87321a52edac9c87bce653a07cf6c2")
+```
+
+which would end up in the YAML as:
+
+```yaml
+- uses: actions/checkout@85e6279cec87321a52edac9c87bce653a07cf6c2 # v4.1.2
+```
+
+Additionally, the bindings server would validate if the version ref (here: `v4.1.2`) points to the mentioned commit
+hash. Thanks to the extra validation, the user an extra assurance than with the YAML approach - the version is for sure
+correct.
+
+Since the version is known and validated, the server can also look up the typing in the catalog.
+
+### 2. Without validation of (version, commit hash) pair
+
+Example:
+
+```kotlin
+@file:Repository("https://bindings.krzeminski.it")
+@file:DependsOn("actions:checkout__commit_lenient:v4.1.2__85e6279cec87321a52edac9c87bce653a07cf6c2")
+```
+
+which also would end up in the YAML as:
+
+```yaml
+- uses: actions/checkout@85e6279cec87321a52edac9c87bce653a07cf6c2 # v4.1.2
+```
+
+It's similar to the above approach, with these differences:
+* the version is just used to add a comment in the YAML - no validation is performed
+* the version is used to look up the typings in the catalog
```

Comment:
This is not really a difference.
With both approaches the typings should be lookupable in the catalog as described if the commit does not have typing information.

Reply:
You're right, I'm not sure why I added it here. Let me remove.